### PR TITLE
[SYCL][ESIMD] Add support for different types for lsc functions

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 1]
```diff
@@ -582,11 +593,20 @@ lsc_block_load(AccessorTy acc, uint32_t offset,
...
   if constexpr (SmallIntFactor == 1) {
-    return __esimd_lsc_load_bti<T, L1H, L3H, _AddressScale, _ImmOffset, _DS,
-                                _VS, _Transposed, N>(pred.data(),
-                                                     offsets.data(), si);
+    if constexpr (_DS == lsc_data_size::u32) {
+      __ESIMD_NS::simd<uint32_t, NElts> result =
+          __esimd_lsc_load_bti<uint32_t, L1H, L3H, _AddressScale, _ImmOffset,
+                               lsc_data_size::u32, _VS, _Transposed, N>(
+              pred.data(), offsets.data(), si);
+      return result.template bit_cast_view<T>();
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:601]: @kbobrovs - accordingly to your experience, is this bit_cast_view() always optimized away if the type is already integral?
If not, then it probably makes sense to write something like:
```
if constexpr (std::is_integral_v<T>) {
    return __esimd_lsc_load_bit<....>(....); 
} else {
    result = __esimd_lsc_load_bit<uint..., >(....);
    return result.template bit_cast_view<T>();
}
```
What is your recommendation here?

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 2]
```diff
@@ -518,9 +519,19 @@ lsc_block_load(const T *p, __ESIMD_NS::simd_mask<1> pred = 1) {
...
   if constexpr (SmallIntFactor == 1) {
-    return __esimd_lsc_load_stateless<T, L1H, L3H, _AddressScale, _ImmOffset,
-                                      _DS, _VS, _Transposed, N>(pred.data(),
-                                                                addrs.data());
+    if constexpr (_DS == lsc_data_size::u32) {
+      __ESIMD_NS::simd<uint32_t, NElts> result =
+          __esimd_lsc_load_stateless<uint32_t, L1H, L3H, _AddressScale,
+                                     _ImmOffset, lsc_data_size::u32, _VS,
+                                     _Transposed, N>(pred.data(), addrs.data());
+      return result.template bit_cast_view<T>();
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:527]: Nit: I suggest to use `sycl::bit_cast` when view is not really needed (like in this case)

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 1 above)
Comment (kbobrovs) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:601]: It is optimized, but I'd prefer `sycl::bit_cast` instead in such cases.

File: sycl/include/sycl/ext/intel/experimental/esimd/common.hpp [hunk 3]
```diff
@@ -185,17 +185,18 @@ constexpr lsc_data_size expand_data_size(lsc_data_size DS) {
 }
 
 template <typename T> struct lsc_expand_type {
-  using type = typename std::conditional<sizeof(T) < 4, uint32_t, T>::type;
+  using type =
+      typename std::conditional<sizeof(T) <= 4, int32_t, int64_t>::type;
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/common.hpp:189]: it fails if only remove 'typename', but it should pass if also add "_t" to std::conditional

**Review by v-klochkov — APPROVED** [review]
Approved taking into account that the functionality is needed and also the accidental/wrong merge of https://github.com/intel/llvm-test-suite/pull/1305
Please though make additional effort trying using sycl::bit_cast instead of bit_cast_view() in follow-up PR.

**Pull request merged**